packages/typegpu/src/core/function/ioOutputType.ts

```diff
@@ -0,0 +1,63 @@
+import { isBuiltin, struct } from '../../data';
+import { type Decorate, attribute, location } from '../../data/attributes';
+import { getCustomLocation, isData } from '../../data/dataTypes';
+import type { BaseWgslData, Location, WgslStruct } from '../../data/wgslTypes';
+import type { FnCore } from './fnCore';
+import type { IOData, IOLayout, Implementation } from './fnTypes';
+
+export type IOLayoutToOutputSchema<T extends IOLayout> = T extends BaseWgslData
+  ? Decorate<T, Location<0>>
+  : T extends Record<string, BaseWgslData>
+    ? WgslStruct<T>
```

Comment:
Wrap every entry of the struct (that isn't a builtin) with a Decorate that adds a `Location<number>`.